src/main/java/io/aiven/kafka/auth/AivenAclAuthorizer.java

```diff
@@ -79,42 +65,52 @@ public void configure(final java.util.Map<String, ?> configs) {
-        checkAndUpdateConfig();
-    }
-
-    /**
-     * Read config file and populate ACL entries, if the config file has changed.
-     * This function assumes appropriate synchronization by caller.
-     */
-    private void checkAndUpdateConfig() {
-        final long configFileLastModified = configFile.lastModified();
-
-        if (configFileLastModified != lastModifiedTimestamp) {
-            LOGGER.info("Reloading ACL configuration {}", configFile);
-            try {
-                final List<AivenAcl> newAclEntries = jsonReader.read();
-                // initialize cache for non-trivial ACLs
-                if (newAclEntries.size() > 10) {
-                    if (verdictCache != null) {
-                        verdictCache.clear();
-                    } else {
-                        verdictCache = new ConcurrentHashMap<>();
-                    }
-                } else {
-                    verdictCache = null;
+        final AclJsonReader jsonReader = new AclJsonReader(configFile.toPath());
+        cacheReference.set(VerdictCache.create(jsonReader));
+        final AtomicReference<WatchKey> watchKeyReference = new AtomicReference<>(subscribeToAclChanges(configFile));
+        scheduledExecutorService.scheduleWithFixedDelay(() -> {
+            final WatchKey watchKey = watchKeyReference.get();
+            if (watchKey != null) {
+                final List<WatchEvent<?>> watchEvents = watchKey.pollEvents();
+
+                watchEvents.stream().filter(watchEvent -> {
+                    @SuppressWarnings("unchecked")
+                    final Path path = ((WatchEvent<Path>) watchEvent).context();
+                    return path.endsWith(configFile.getName());
+                }).findFirst().ifPresent(watchEvent -> {
+                    LOGGER.info(watchEvent.kind() + ": " + watchEvent.context()
+                            + ", Modified: " + configFile.lastModified());
+                    cacheReference.set(VerdictCache.create(jsonReader));
+                });
+                if (!watchKey.reset()) {
+                    watchKeyReference.compareAndSet(watchKey, null);
                 }
-                aclEntries = newAclEntries;
-            } catch (final JsonReaderException ex) {
-                LOGGER.error("Failed to load config file", ex);
+            } else {
+                watchKeyReference.set(subscribeToAclChanges(configFile));
             }
+        }, 0, config.configRefreshTimeout(), TimeUnit.MILLISECONDS);
+    }
+
+    private WatchKey subscribeToAclChanges(final File configFile) {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            return configFile.toPath().getParent()
+                    .register(watchService, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_DELETE);
+        } catch (final IOException e) {
+            LOGGER.error("Failed to subscribe to ACL configuration changes");
```

Comment:
A suggestion from a composite character ops engineer, reading this logs at 23:40 :)
```suggestion
            LOGGER.error("Failed to subscribe to ACL configuration changes", e);
```